### PR TITLE
feat(plasma-ui): add support of @salutejs/spatial, fix context memoization

### PR DESCRIPTION
**Pull request by arsenykruglikov**
Button ✅
- Confirm ✅ 
- Stepper ✅ 

Input ✅
- Checkbox общий ✅ 
- Radiobox общий ✅ 
- TextField собственный ✅ 
- Switch собственный ✅ 

TextArea ✅ 

Picker: ✅

Carousel: нет смысла делать навигацию на CarouselItem так как внутри него может лежать что угодно, в том числе что-то для навигации. И во-вторых к CarouselItem и CarouseCol можно применить className самостоятельно.

PaginationDots: нет интерактива

Slider: не понятно как передвигать react-draggabe компонент с помощью стрелочек на клавиатуре

Tabs: Используется в сочетании TabItem на который класс можно проставить самостоятельно

Toast: нет интерактива

:baby_chick: Download canary assets:

[colors--canary.285.4001445129.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.285.4001445129.xml)  
[color_metro_ios-swift--canary.285.4001445129.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_ios-swift--canary.285.4001445129.swift)  
[color_metro_kotlin--canary.285.4001445129.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_kotlin--canary.285.4001445129.kt)  
[color_metro_react-native--canary.285.4001445129.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_react-native--canary.285.4001445129.ts)  
[color_sbermarket_ios-swift--canary.285.4001445129.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.285.4001445129.swift)  
[color_sbermarket_kotlin--canary.285.4001445129.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.285.4001445129.kt)  
[color_sbermarket_react-native--canary.285.4001445129.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.285.4001445129.ts)  
[color_sberprime_ios-swift--canary.285.4001445129.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.285.4001445129.swift)  
[color_sberprime_kotlin--canary.285.4001445129.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.285.4001445129.kt)  
[color_sberprime_react-native--canary.285.4001445129.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.285.4001445129.ts)  
[color_selgros_ios-swift--canary.285.4001445129.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_ios-swift--canary.285.4001445129.swift)  
[color_selgros_kotlin--canary.285.4001445129.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_kotlin--canary.285.4001445129.kt)  
[color_selgros_react-native--canary.285.4001445129.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_react-native--canary.285.4001445129.ts)  
[color_smbusiness_ios-swift--canary.285.4001445129.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_ios-swift--canary.285.4001445129.swift)  
[color_smbusiness_kotlin--canary.285.4001445129.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_kotlin--canary.285.4001445129.kt)  
[color_smbusiness_react-native--canary.285.4001445129.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_react-native--canary.285.4001445129.ts)  
[PlasmaTokensColor--canary.285.4001445129.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.285.4001445129.swift)  
[typo_mage_ios-swift--canary.285.4001445129.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_ios-swift--canary.285.4001445129.swift)  
[typo_mage_kotlin--canary.285.4001445129.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.285.4001445129.kt)  
[typo_mage_react-native--canary.285.4001445129.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.285.4001445129.ts)  
[typo_plasma_ios-swift--canary.285.4001445129.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_ios-swift--canary.285.4001445129.swift)  
[typo_plasma_kotlin--canary.285.4001445129.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.285.4001445129.kt)  
[typo_plasma_react-native--canary.285.4001445129.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.285.4001445129.ts)  
[typo_ruler_ios-swift--canary.285.4001445129.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_ios-swift--canary.285.4001445129.swift)  
[typo_ruler_kotlin--canary.285.4001445129.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.285.4001445129.kt)  
[typo_ruler_react-native--canary.285.4001445129.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.285.4001445129.ts)  
[typo_sage_ios-swift--canary.285.4001445129.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_ios-swift--canary.285.4001445129.swift)  
[typo_sage_kotlin--canary.285.4001445129.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.285.4001445129.kt)  
[typo_sage_react-native--canary.285.4001445129.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.285.4001445129.ts)  
[typo_sbermarket_ios-swift--canary.285.4001445129.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_ios-swift--canary.285.4001445129.swift)  
[typo_sbermarket_kotlin--canary.285.4001445129.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.285.4001445129.kt)  
[typo_sbermarket_react-native--canary.285.4001445129.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.285.4001445129.ts)  
[typo_soulmate_ios-swift--canary.285.4001445129.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_ios-swift--canary.285.4001445129.swift)  
[typo_soulmate_kotlin--canary.285.4001445129.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.285.4001445129.kt)  
[typo_soulmate_react-native--canary.285.4001445129.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.285.4001445129.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-b2c@1.127.0-canary.285.4001445129.0
  npm install @salutejs/plasma-core@1.90.0-canary.285.4001445129.0
  npm install @salutejs/plasma-icons@1.118.0-canary.285.4001445129.0
  npm install @salutejs/plasma-temple@1.125.0-canary.285.4001445129.0
  npm install @salutejs/plasma-ui@1.158.0-canary.285.4001445129.0
  npm install @salutejs/plasma-web@1.154.0-canary.285.4001445129.0
  npm install @salutejs/plasma-sb-utils@0.88.0-canary.285.4001445129.0
  # or 
  yarn add @salutejs/plasma-b2c@1.127.0-canary.285.4001445129.0
  yarn add @salutejs/plasma-core@1.90.0-canary.285.4001445129.0
  yarn add @salutejs/plasma-icons@1.118.0-canary.285.4001445129.0
  yarn add @salutejs/plasma-temple@1.125.0-canary.285.4001445129.0
  yarn add @salutejs/plasma-ui@1.158.0-canary.285.4001445129.0
  yarn add @salutejs/plasma-web@1.154.0-canary.285.4001445129.0
  yarn add @salutejs/plasma-sb-utils@0.88.0-canary.285.4001445129.0
  ```
</details>

### CODE REVIEW

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🌚 Missing base report**

Check out current report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-285/

File: packages/plasma-core/src/components/Basebox/Basebox.tsx [hunk 1]
```diff
@@ -41,7 +41,13 @@ export const StyledContentWrapper = styled.label`
     cursor: inherit;
 `;
 
-export const StyledInput = styled.input`
+const SpatnavInput = (props: ComponentPropsWithRef<'input'>) => {
+    props.className = `${props.className} sn-section-item`;
```
Comment (Yeti-or) [inline packages/plasma-core/src/components/Basebox/Basebox.tsx:45]: а почему тут не использовать attrs ? 

https://styled-components.com/docs/basics#attaching-additional-props

File: packages/plasma-core/src/components/Button/Button.tsx [hunk 2]
```diff
@@ -1,16 +1,23 @@
...
+import React, { ComponentPropsWithRef } from 'react';
 import styled from 'styled-components';
 
 import { applyDisabled, applyBlur, applyEllipsis } from '../../mixins';
 import { Spinner } from '../Spinner';
 
 import type { ButtonIsContentProps, ButtonIsLoading, StyledButtonProps } from './Button.types';
 
+const SpatnavButton = (props: ComponentPropsWithRef<'button'>) => {
+    props.className = `${props.className} sn-section-item`;
```
Comment (Yeti-or) [inline packages/plasma-core/src/components/Button/Button.tsx:10]: `sn-section-item` – было бы здорово  коммент c ссылкой на доку spatnav

потому что `sn` не прям говорящее название если не знаешь про sn

File: packages/plasma-ui/src/components/Tabs/TabItem.tsx [hunk 3]
```diff
@@ -82,18 +82,20 @@ export const StyledTabItemMemo = React.memo(StyledTabItem);
 /**
  * Элемент списка вкладок, недопустимо использовать вне компонента Tabs.
  */
-/**
- * Элемент списка вкладо`к, недопустимо использовать вне компонента Tabs.
- */
```
Comment (Yeti-or) [inline packages/plasma-ui/src/components/Tabs/TabItem.tsx:87]: 👍

**Review by Yeti-or — COMMENTED** [review]
А почему классы добавляются в core а не в ui ?

**Comment by Yeti-or** [comment]
@arsenykruglikov что-то всё развалолось в pr

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-285/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🌚 Missing base report**

Check out current report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-285/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

File: packages/plasma-core/src/components/Basebox/Basebox.tsx (hunk 1 above)
Comment (arsenykruglikov) [reply packages/plasma-core/src/components/Basebox/Basebox.tsx]: Мне показалось сначала, что так не очень, но я вынес колбек в отдельную функцию и получилось нормально

File: packages/plasma-core/src/components/Button/Button.tsx (hunk 2 above)
Comment (arsenykruglikov) [reply packages/plasma-core/src/components/Button/Button.tsx]: Добавил коммент в функцию добавления className

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-285/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-285/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-285/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-285/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-285/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-285/temple-storybook/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-285/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-285/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-285/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-285/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-285/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-285/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-285/temple-storybook/

**Review by Yakutoc — APPROVED** [review]

**Comment by Yakutoc** [comment]
@arsenykruglikov Привет, объясни пожалуйста что именно вызвало трудности? ) 
   
> Slider: не понятно как передвигать react-draggable компонент с помощью стрелочек на клавиатуре

**Review by neretin-trike — APPROVED** [review]

File: packages/plasma-ui/src/components/Tabs/TabsAnimationContext.tsx [hunk 4]
```diff
@@ -15,10 +15,6 @@ export class TabItemRefs {
     }
 }
 
-interface TabsState {
-    refs?: TabItemRefs;
```
Comment (Yeti-or) [inline packages/plasma-ui/src/components/Tabs/TabsAnimationContext.tsx:19]: это чтобы не создавать лишних объектов?

**Comment by Yeti-or** [comment]
@arsenykruglikov я всё ещё не могу понять почему эти изменения в  plasma-core а не в plasma-ui ?

**Review by Yeti-or — CHANGES_REQUESTED** [review]

**Review by Yeti-or — CHANGES_REQUESTED** [review]
@arsenykruglikov я всё ещё не могу понять почему эти изменения в plasma-core а не в plasma-ui ?

File: packages/plasma-ui/src/components/Tabs/TabsAnimationContext.tsx (hunk 4 above)
Comment (arsenykruglikov) [reply packages/plasma-ui/src/components/Tabs/TabsAnimationContext.tsx:19]: И объект не создаем и в useMemo нет нагромождения

**Comment by arsenykruglikov** [comment]
@Yeti-or на счёт того, почему классы добавляются в `plasma-core`. Из-за того, что `plasma-ui` намертво завязана на оригинальные компоненты из `core`  и переиспользует в разных местах. Эти классы не влияют на стили, поэтому я не увидел препятствий, чтобы сделать именно так как сделано

**Comment by Yeti-or** [comment]
@arsenykruglikov ну меня смущает что в plasma-web будет приезжать лишний код, я не совсем понял , в компонентах plasma-ui не получается докинуть классы? или что значит намертво

**Comment by arsenykruglikov** [comment]
@Yeti-or 
> в компонентах plasma-ui не получается докинуть классы?

Можно, конечно, но придётся следить за тем, что в каждое использование соответствующих компонентов из `core` добавлен класс. А что если кто-то добавит новый тип кнопки и забудет проставить класс.

> или что значит намертво

Это значит, что пакет `ui` использует пакет `core` как основу и берёт базовые компоненты оттуда, иного даже просто делает реэкспорт даже не изменяя ничего

**Comment by arsenykruglikov** [comment]
@Yeti-or @neretin-trike я переписал так, чтобы классы добавлялись только в plasma-ui

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-285/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🌚 Missing base report**

Check out current report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-285/

**Comment by neretin-trike** [comment]
@arsenykruglikov По поводу Picker обсудили. Пришли к тому, что фокусироваться надо только на те "барабаны", что в середине

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-285/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🔴 Fail**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🔴 Fail**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-285/

File: packages/plasma-ui/src/helpers/spatnavClassNameAttrs.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Добавляет класс `sn-section-item`, необходимый для работы @salutejs/spatial
+ */
+export const spatnavClassNameAttrs: { className?: string } = {
```
Comment (Yeti-or) [inline packages/plasma-ui/src/helpers/spatnavClassNameAttrs.ts:4]: можешь только не в helpers а в utils положить – прости это наша не описанная папка, мы в helpers кладем все про stotybook =(

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🔴 Fail**

Check out comparison report in job artifacts!

File: packages/plasma-ui/src/components/Button/Button.tsx [hunk 6]
```diff
@@ -36,7 +36,10 @@ const ButtonBase = createButton<HTMLButtonElement, ButtonProps & ThemeProviderCo
 export const Button = forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
     const { deviceScale } = useThemeContext() || {};
 
-    return <ButtonBase {...props} ref={ref} deviceScale={deviceScale} />;
+    // Добавляем класс `sn-section-item`, необходимый для работы @salutejs/spatial
+    return (
```
Comment (Yeti-or) [inline packages/plasma-ui/src/components/Button/Button.tsx:40]: кстати тут не нужно:

```
const { className, ...res } = props;

return (
  <ButtonBase {...res} ref={ref} className={`${className} sn-section-item`} deviceScale={deviceScale} />
);
```

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-285/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-285/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-285/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-285/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-285/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-285/temple-storybook/

**Comment by Yeti-or** [comment]
в Slider / Picker в итоге не нужны такие классы?

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-285/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-285/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🔴 Fail**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-285/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-285/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-285/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-285/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-285/temple-storybook/

**Comment by arsenykruglikov** [comment]
@Yeti-or вроде бы всё, можно финально смотреть

**Review by Yeti-or — APPROVED** [review]

**Pull request merged**